torch/jit/__init__.py

```diff
@@ -1049,9 +1050,33 @@ def _compile_and_register_class(obj, rcb, qualified_name):
     _add_script_class(obj, qualified_name)
 
 
+def _compile_function(fn, qualified_name, _rcb=None, _frames_up=1):
+    ast = get_jit_def(fn)
+    if _rcb is None:
+        closure_rcb = _jit_internal.createResolutionCallbackFromClosure(fn)
+        stack_rcb = _jit_internal.createResolutionCallback(_frames_up + 1)
+
+        def _rcb(name):
+            # since type comments aren't captured in the function's closures,
+            # we still need to try to the rcb based on stack frames if the
+            # closure rcb fails
+            result = closure_rcb(name)
+            if result:
+                return result
+            return stack_rcb(name)
+    fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(fn))
+    # Forward docstrings
+    fn.__doc__ = fn.__doc__
+    return fn
+
+
 def script(obj, optimize=None, _frames_up=0, _rcb=None):
     if not _enabled:
         return obj
+
+    # In case there were some previous failed compilations, clear out the stack
+    torch._C._clear_compilation_error_stack()
```

Comment:
I'm suspicious of using manual calls to manipulate the error stack state. Can we not make pushing/popping an RAII-like property of the IR generation step? so the constructor/destructor of `to_ir` pushes/pops off the stack.

Reply:
That would be good but I don't know if it can work here. With changing `CallStack`'s constructor / destructor to replace `push/pop_function`, it looks like the stack-unwinding happens before the `exception::what()` is called, so the stack gets deleted before it's shown.